File: orion-ui/types/savedArtifact.ts
/* -------------------------------------------------------------------------- */
/* Saved Artifacts                                                            */
/* -------------------------------------------------------------------------- */

/**

* Where a saved artifact is stored/displayed.
  */
  export type SaveDestination =
  | "library"
  | "research"
  | "reports";

/* -------------------------------------------------------------------------- */
/* Saved Artifact                                                             */
/* -------------------------------------------------------------------------- */

/**

* Canonical frontend representation of a saved artifact.
*
* The backend currently uses integer research IDs in PostgreSQL, while the
* frontend/API layer also supports UUID/string research identifiers.
*
* Therefore both snake_case API fields and the normalized camelCase research
* identifier are supported at the type boundary.
  */
  export interface SavedArtifact {
  id: string;

/**

* Backend/API research identifier.
*
* Supports:
* * PostgreSQL integer IDs
* * UUID/string IDs
    */
    research_id: string | number;

/**

* Normalized frontend research identifier.
*
* API mappers may expose this field for UI components that use camelCase.
  */
  researchId?: string | number | null;

destination: SaveDestination;

title: string;

description?: string | null;

created_at: string;

updated_at?: string | null;
}

/* -------------------------------------------------------------------------- */
/* Create / Save Request                                                      */
/* -------------------------------------------------------------------------- */

/**

* Payload sent to the backend when saving an artifact.
  */
  export interface SaveArtifactRequest {
  /**

  * Research identifier.
* * The API client accepts both numeric and string/UUID identifiers.
    */
    research_id: string | number;

destination: SaveDestination;

title: string;

description?: string | null;
}

/* -------------------------------------------------------------------------- */
/* API Response                                                               */
/* -------------------------------------------------------------------------- */

/**

* Response returned by POST /api/saved-artifacts.
*
* The backend may currently return either:
*
* 1. A wrapped response:
*
* {
* ```
   "success": true,
  ```
* ```
   "artifact": { ... }
  ```
* }
*
* 2. The SavedArtifact directly:
*
* {
* ```
   "id": "...",
  ```
* ```
   "research_id": 73,
  ```
* ```
   ...
  ```
* }
*
* The API layer normalizes either form before exposing the result to the
* application.
  */
  export interface SaveArtifactResponse {
  success?: boolean;

artifact?: SavedArtifact;

/**

* Direct SavedArtifact response fields.
  */
  id?: string;

research_id?: string | number;

destination?: SaveDestination;

title?: string;

description?: string | null;

created_at?: string;

updated_at?: string | null;
}

File: backend/app/schemas/saved_artifact.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.db.models.saved_artifact import SaveDestination


class SavedArtifactCreate(BaseModel):
    """
    Request body for creating a saved artifact.

    Research.id is a UUID, so research_id must be a UUID.
    """

    research_id: UUID

    destination: SaveDestination

    title: str = Field(
        ...,
        min_length=1,
        max_length=500,
    )

    description: str | None = None


class SavedArtifactResponse(BaseModel):
    """
    Response returned for a saved artifact.
    """

    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID

    research_id: UUID

    destination: SaveDestination

    title: str

    description: str | None

    created_at: datetime

    updated_at: datetime | None = None


class SavedArtifactListResponse(BaseModel):
    """
    Response returned when listing saved artifacts.
    """

    items: list[SavedArtifactResponse]

    total: int